Differential convolution

mliuschi opened the pull request:
Pull request for adding differential convolution. The current implementation uses finite difference convolutions as parallel branches with FNO, just like the skip connections. As such, the main changes are in `FNOBlocks`.

dhpitt commented:
Thanks for the PR @mliuschi! The new finite difference conv file looks good, just some comments about the best way to implement. We should also unit-test the new layer before merging this PR.

mliuschi commented on neuralop/layers/differential_conv.py:
```diff
@@ -0,0 +1,89 @@
+            conv_mean = conv_mean.repeat([1 for i in range(len(conv_mean.shape) - self.num_dim)] + [self.kernel_size for j in range(self.num_dim)])
+            conv_x = self.F_conv_module(input=x_pad, weight=(self.conv_kernel - conv_mean), padding='valid', groups=self.groups)
+            return conv_x / grid_width
+        else:
```
I think there's already a `NotImplementedError` in the init function (line 74). I guess it is redundant at line 89 then. Would you recommend I remove it and replace line 82 with a generic `else`? I didn't do this initially for increased readability so it's easier to see which implementation is being used in the `forward` function. It also might be clearer if more implementations happen to be added later.

JeanKossaifi replied:
I agree it's clearer to have the test done at __init__ and not wait for forward pass.

An option is to define multiple forward passes (forward_substract_middle, etc) and just use the correct one. Might also be more readible.

A (perhaps better?) option is to have a function that applies each of these, with a unit-test (like in torch.nn.functional) and we just use them here.

mliuschi commented:
I've added two tests:
1. Test of the full `LocalFNOBlocks` with varying parameters, just like in `FNOBlocks`. The output scaling factor is also tested.
2. Test of convergence for `FiniteDifferenceConvolution`. This was performed in the paper, and I've added it here because I think it's a useful sanity check.

dhpitt reviewed (approved):
These changes look great. Thanks @mliuschi ! Requesting review from @JeanKossaifi

JeanKossaifi commented on neuralop/layers/tests/test_finite_difference_conv.py:
```diff
@@ -0,0 +1,72 @@
+
+    with torch.no_grad():
+        if implementation == 'subtract_middle':
+            weight = differential_block.conv.weight[0] # [0] because output channels is 1
```
MIght be worse separating the unit-tests for each type (substract middle, all) so it is also easier to debug and add more later.

mliuschi commented:
@JeanKossaifi I've made some style changes to make things cleaner. I've also defined the different forward passes for each implementation for now. Let me know what you think! If you think it's best to write "torch.nn.functional-style" operations, I can do that too!

JeanKossaifi commented:
Looks good, thanks @mliuschi !
